sklearn/metrics/cluster/stability.py

```diff
@@ -0,0 +1,157 @@
+    prop_subset: 0 < float < 1: proportion of input data taken in each subset
+    distance: a string naming a distance or a cluster similarity. can be in
+        ['euclidian', 'minkowski', 'seuclidiean', 'sqeuclidean', 'chebyshev'
+         'cityblock', 'cosine', 'correlation', 'hamming', 'jaccard',
+         'Bray-Curtis', 'mahalanobis', 'yule', 'matching', 'dice', 'kulsinski',
+         'rogerstanimoto', 'russellrao', 'sokalmichener', 'sokalsneath',
+         'canberra', 'wminkowski', 'fowlkes-mallows'])
+    p : double
+        The p-norm to apply (for Minkowski, weighted and unweighted)
+    Return
+    ------
+    k: int
+    """
+    rng = check_random_state(random_state)
+    cluster_similarity = function_cluster_similarity(distance, p)
+
+    n_samples, n_features = X.shape
+    if not k_max:
+        k_max = n_samples // 2
+
+    best_stab, best_k = 0, 0
+    for k in range(2, k_max + 1):
+        cluster_estimator.set_params(n_clusters=k)
+        this_score = sum(
+            _one_stability_measure(cluster_estimator, X, prop_subset,
+                                   cluster_similarity)
+            for _ in range(nb_draw)) / nb_draw
+        if verbose:
+            print('for %d cluster, stability is %f' % (k, this_score))
+
+        if this_score >= best_stab:
+            best_stab = this_score
+            best_k = k
+
+    return best_k
+
+
+def _one_stability_measure(cluster_estimator, X, prop_sample,
+                           cluster_similarity, random_state=None):
+    """
+    Draws two subsets A and B from X, compute C_A, clustering on subset
+    A, and C_B, clustering on subset B, then returns
+
+    similarity(C_A, C_B)
+
+    Parameter
+    ---------
+    X: array of size n_samples, n_features
+    cluster_estimator: ClusterMixing estimator object.
+        need parameter n_clusters
+        need method fit_predict: X -> labels
+    prop_sample: 0 < float < 1, proportion of X taken in each subset
+    cluster_similarity: function (list, list) -> float
+    """
+    rng = check_random_state(random_state)
+
+    n_sample = X.shape[0]
+    set_1 = rng.uniform(size=n_sample) < prop_sample
+    set_2 = rng.uniform(size=n_sample) < prop_sample
+    nb_points_1, nb_points_2 = 0, 0
+    points_1, points_2 = [], []
+    common_points_1, common_points_2 = [], []
+    for i, (is_1, is_2) in enumerate(zip(set_1, set_2)):
+        if is_1 and is_2:
+            common_points_1.append(nb_points_1)
+            common_points_2.append(nb_points_2)
+        if is_1:
+            points_1.append(i)
+            nb_points_1 += 1
+        if is_2:
+            points_2.append(i)
+            nb_points_2 += 1
+
+    assi_1 = cluster_estimator.fit_predict(X[np.ix_(points_1)])
+    assi_2 = cluster_estimator.fit_predict(X[np.ix_(points_2)])
+
+    clustering_1 = [assi_1[c] for c in common_points_1]
+    clustering_2 = [assi_2[c] for c in common_points_2]
+    return cluster_similarity(clustering_1, clustering_2)
```

Comment:
If the order of the clusters changes, then I think all metrics besides 'fowlkes-mallows' will be computed incorrectly?